File: scripts/integrations/tailscale.py
#!/usr/bin/env python3
"""
Cortex Tailscale integration — mesh network connector.

Brings up Tailscale headlessly using an auth key stored in the Cortex vault.
Makes home network resources (NAS, SFTP, SMB) reachable from anywhere —
enabling rclone local/SFTP backends to work on mobile, Chromebook, or laptop
away from home.

Mobile devices use the Tailscale app directly — this connector is for
Linux / macOS / Windows CLI environments (steve-cachyos, Chromebook, laptop).

Setup:
  python scripts/integrations/tailscale.py auth

Usage:
  python scripts/integrations/tailscale.py auth              [--passphrase <p>]
  python scripts/integrations/tailscale.py status            [--passphrase <p>]
  python scripts/integrations/tailscale.py up                [--passphrase <p>]
  python scripts/integrations/tailscale.py down
  python scripts/integrations/tailscale.py ip <hostname>     [--passphrase <p>]
  python scripts/integrations/tailscale.py peers             [--passphrase <p>]

Requires: tailscale installed and in PATH
  Linux:   curl -fsSL https://tailscale.com/install.sh | sh
  macOS:   brew install tailscale
  ChromeOS (Crostini): curl -fsSL https://tailscale.com/install.sh | sh
"""

import os
import sys
import json
import argparse
import subprocess

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

VAULT_KEY = "tailscale-auth-key"


# ── tailscale check ───────────────────────────────────────────────────────────

def require_tailscale():
    if subprocess.run(["which", "tailscale"], capture_output=True).returncode != 0:
        print("ERROR: tailscale is not installed.")
        print("  Linux / ChromeOS: curl -fsSL https://tailscale.com/install.sh | sh")
        print("  macOS:            brew install tailscale")
        sys.exit(1)


# ── Vault helpers ─────────────────────────────────────────────────────────────

def get_secret(name: str, passphrase: str) -> str:
    secrets_script = os.path.join(ROOT, "scripts", "secrets.py")
    result = subprocess.run(
        [sys.executable, secrets_script, "get", name, "--passphrase", passphrase],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        print(f"ERROR: Could not retrieve '{name}' from vault.")
        print("Run: python scripts/integrations/tailscale.py auth")
        sys.exit(1)
    return result.stdout.strip()


def store_secret(name: str, value: str, passphrase: str):
    secrets_script = os.path.join(ROOT, "scripts", "secrets.py")
    subprocess.run(
        [sys.executable, secrets_script, "store", name,
         "--value", value, "--passphrase", passphrase],
        check=True
    )


# ── Auth ──────────────────────────────────────────────────────────────────────

def cmd_auth(passphrase: str):
    """Store a Tailscale auth key in the vault."""
    print("Tailscale auth key setup")
    print("Generate a reusable auth key at: https://login.tailscale.com/admin/settings/keys")
    print("Settings: Reusable ON, Ephemeral OFF\n")

    key = input("Paste auth key: ").strip()
    if not key.startswith("tskey-"):
        print("WARNING: key doesn't look like a Tailscale auth key (expected tskey-...)")

    store_secret(VAULT_KEY, key, passphrase)
    print(f"\nStored as '{VAULT_KEY}' in vault.")
    print("Run: python scripts/integrations/tailscale.py up")


# ── Status ────────────────────────────────────────────────────────────────────

def cmd_status():
    require_tailscale()
    result = subprocess.run(["tailscale", "status"], capture_output=True, text=True)
    if result.returncode != 0:
        print("Tailscale is not running or not connected.")
        print("Run: python scripts/integrations/tailscale.py up")
        return
    print(result.stdout)


# ── Up ────────────────────────────────────────────────────────────────────────

def cmd_up(passphrase: str):
    require_tailscale()
    auth_key = get_secret(VAULT_KEY, passphrase)

    print("Bringing Tailscale up...")
    result = subprocess.run(
        ["tailscale", "up", "--authkey", auth_key],
        capture_output=True, text=True
    )

    if result.returncode == 0:
        print("Tailscale is up.")
        cmd_status()
    else:
        print(f"ERROR: {result.stderr.strip()}")
        sys.exit(1)


# ── Down ──────────────────────────────────────────────────────────────────────

def cmd_down():
    require_tailscale()
    result = subprocess.run(["tailscale", "down"], capture_output=True, text=True)
    if result.returncode == 0:
        print("Tailscale is down.")
    else:
        print(f"ERROR: {result.stderr.strip()}")
        sys.exit(1)


# ── IP ────────────────────────────────────────────────────────────────────────

def cmd_ip(hostname: str):
    """Get the Tailscale IP of a peer by hostname. Use this in rclone SFTP config."""
    require_tailscale()
    result = subprocess.run(
        ["tailscale", "ip", hostname],
        capture_output=True, text=True
    )
    if result.returncode == 0:
        ip = result.stdout.strip()
        print(f"{hostname}: {ip}")
        print(f"\nUse this IP in your rclone SFTP remote config:")
        print(f"  host = {ip}")
    else:
        print(f"ERROR: {result.stderr.strip()}")
        print("Is Tailscale up? Run: python scripts/integrations/tailscale.py up")
        sys.exit(1)


# ── Peers ─────────────────────────────────────────────────────────────────────

def cmd_peers():
    """List all peers in the tailnet with their IPs and hostnames."""
    require_tailscale()
    result = subprocess.run(
        ["tailscale", "status", "--json"],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        print("Tailscale is not running. Run: python scripts/integrations/tailscale.py up")
        sys.exit(1)

    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError:
        print("ERROR: Could not parse tailscale status output.")
        sys.exit(1)

    peers = data.get("Peer", {})
    self_node = data.get("Self", {})

    print("# Tailnet peers\n")

    # Self
    self_ips = self_node.get("TailscaleIPs", [])
    self_host = self_node.get("HostName", "this device")
    self_ip = self_ips[0] if self_ips else "unknown"
    print(f"  {self_ip}  {self_host}  (this device)")

    # Peers
    for _, peer in peers.items():
        ips = peer.get("TailscaleIPs", [])
        ip = ips[0] if ips else "unknown"
        hostname = peer.get("HostName", "unknown")
        online = "online" if peer.get("Online") else "offline"
        os_name = peer.get("OS", "")
        os_str = f" [{os_name}]" if os_name else ""
        print(f"  {ip}  {hostname}{os_str}  ({online})")

    print(f"\nTo get IP for rclone: python scripts/integrations/tailscale.py ip <hostname>")


# ── Main ──────────────────────────────────────────────────────────────────────

def prompt_passphrase() -> str:
    import getpass
    return getpass.getpass("Vault passphrase: ")


def main():
    parser = argparse.ArgumentParser(
        description="Cortex Tailscale integration — mesh network connector"
    )
    parser.add_argument("--passphrase", default=None)
    sub = parser.add_subparsers(dest="cmd")

    sub.add_parser("auth", help="Store Tailscale auth key in vault")
    sub.add_parser("status", help="Show Tailscale connection status")
    sub.add_parser("up", help="Bring Tailscale up using vaulted auth key")
    sub.add_parser("down", help="Bring Tailscale down")

    p_ip = sub.add_parser("ip", help="Get Tailscale IP of a peer (use in rclone config)")
    p_ip.add_argument("hostname", help="Peer hostname (e.g. steve-cachyos, nas)")

    sub.add_parser("peers", help="List all peers in the tailnet")

    args = parser.parse_args()

    if not args.cmd:
        parser.print_help()
        sys.exit(1)

    # Commands that don't need the vault
    if args.cmd == "status":
        cmd_status()
        return
    if args.cmd == "down":
        cmd_down()
        return
    if args.cmd == "peers":
        cmd_peers()
        return
    if args.cmd == "ip":
        cmd_ip(args.hostname)
        return

    # Commands that need the vault
    passphrase = args.passphrase or prompt_passphrase()

    if args.cmd == "auth":
        cmd_auth(passphrase)
    elif args.cmd == "up":
        cmd_up(passphrase)


if __name__ == "__main__":
    main()

File: manifest/framework/scripts/integrations/tailscale.ts
#!/usr/bin/env bun
/**
 * Cortex Tailscale integration — mesh network connector.
 *
 * Brings up Tailscale headlessly using an auth key stored in the Cortex vault.
 *
 * Setup:
 *   bun manifest/framework/scripts/integrations/tailscale.ts auth
 *
 * Usage:
 *   bun manifest/framework/scripts/integrations/tailscale.ts [--passphrase <p>] auth
 *   bun manifest/framework/scripts/integrations/tailscale.ts status
 *   bun manifest/framework/scripts/integrations/tailscale.ts [--passphrase <p>] up
 *   bun manifest/framework/scripts/integrations/tailscale.ts down
 *   bun manifest/framework/scripts/integrations/tailscale.ts ip <hostname>
 *   bun manifest/framework/scripts/integrations/tailscale.ts peers
 *
 * Requires: tailscale installed and in PATH
 */

import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import { tmpdir } from 'node:os'
import { writeFileSync, unlinkSync } from 'node:fs'
import * as readline from 'node:readline'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '../../../..')
const VAULT_KEY = 'tailscale-auth-key'

// --- Arg parser ---

function parseArgs(argv: string[]): Record<string, string | boolean | undefined> {
  const result: Record<string, string | boolean | undefined> = {}
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]
    if (arg.startsWith('--')) {
      const key = arg.slice(2)
      const next = argv[i + 1]
      if (next && !next.startsWith('--')) {
        result[key] = next
        i++
      } else {
        result[key] = true
      }
    } else if (!result['_cmd']) {
      result['_cmd'] = arg
    } else if (!result['_pos1']) {
      result['_pos1'] = arg
    }
  }
  return result
}

// --- Interactive prompts ---

async function promptLine(label: string): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error(`ERROR: ${label} required but stdin is not a TTY.`)
    process.exit(1)
  }
  return new Promise(resolve => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout })
    rl.question(`${label}: `, answer => {
      rl.close()
      resolve(answer.trim())
    })
  })
}

async function promptPassphrase(label = 'Vault passphrase'): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error('ERROR: No passphrase provided and stdin is not a TTY. Use --passphrase.')
    process.exit(1)
  }
  return new Promise(resolve => {
    process.stderr.write(`${label}: `)
    let value = ''
    const tty = process.stdin as NodeJS.ReadStream & { setRawMode?: (mode: boolean) => void }
    if (tty.setRawMode) {
      process.stdin.setRawMode(true)
      process.stdin.resume()
      process.stdin.setEncoding('utf8')
      const onData = (ch: string) => {
        if (ch === '\r' || ch === '\n') {
          process.stdin.setRawMode!(false)
          process.stdin.pause()
          process.stdin.removeListener('data', onData)
          process.stderr.write('\n')
          resolve(value)
        } else if (ch === '') {
          process.exit(1)
        } else if (ch === '') {
          value = value.slice(0, -1)
        } else {
          value += ch
        }
      }
      process.stdin.on('data', onData)
    } else {
      const rl = readline.createInterface({ input: process.stdin, output: process.stderr })
      rl.question('', answer => { rl.close(); resolve(answer) })
    }
  })
}

// --- tailscale check ---

function requireTailscale(): void {
  const proc = Bun.spawnSync(['which', 'tailscale'], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode !== 0) {
    console.error('ERROR: tailscale is not installed.')
    console.error('  Linux / ChromeOS: curl -fsSL https://tailscale.com/install.sh | sh')
    console.error('  macOS:            brew install tailscale')
    process.exit(1)
  }
}

// --- Vault helpers ---

async function getSecret(name: string, passphrase: string): Promise<string> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'get', name, '--passphrase', passphrase],
    { stdout: 'pipe', stderr: 'pipe' }
  )
  const code = await proc.exited
  if (code !== 0) {
    console.error(`ERROR: Could not retrieve '${name}' from vault.`)
    console.error('Run: bun manifest/framework/scripts/integrations/tailscale.ts auth')
    process.exit(1)
  }
  return (await new Response(proc.stdout).text()).trim()
}

async function storeSecret(name: string, value: string, passphrase: string): Promise<void> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'store', name, '--value', value, '--passphrase', passphrase],
    { stdout: 'inherit', stderr: 'inherit' }
  )
  const code = await proc.exited
  if (code !== 0) {
    console.error(`ERROR: Failed to store secret '${name}'.`)
    process.exit(1)
  }
}

// --- Auth ---

async function cmdAuth(passphrase: string): Promise<void> {
  console.log('Tailscale auth key setup')
  console.log('Generate a reusable auth key at: https://login.tailscale.com/admin/settings/keys')
  console.log('Settings: Reusable ON, Ephemeral OFF\n')

  const key = await promptLine('Paste auth key')
  if (!key.startsWith('tskey-')) {
    console.log("WARNING: key doesn't look like a Tailscale auth key (expected tskey-...)")
  }

  await storeSecret(VAULT_KEY, key, passphrase)
  console.log(`\nStored as '${VAULT_KEY}' in vault.`)
  console.log('Run: bun manifest/framework/scripts/integrations/tailscale.ts up')
}

// --- Status ---

function cmdStatus(): void {
  requireTailscale()
  const proc = Bun.spawnSync(['tailscale', 'status'], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode !== 0) {
    console.log('Tailscale is not running or not connected.')
    console.log('Run: bun manifest/framework/scripts/integrations/tailscale.ts up')
    return
  }
  console.log(proc.stdout.toString())
}

// --- Up ---

async function cmdUp(passphrase: string): Promise<void> {
  requireTailscale()
  const authKey = await getSecret(VAULT_KEY, passphrase)

  // Passing the key as `--authkey <value>` exposes it in the tailscale
  // process's argv — readable via `ps` / /proc/<pid>/cmdline by any local
  // user for the life of the call. Hand it over through a mode-0600 temp file
  // and tailscale's `file:` form instead, and scrub the file even on interrupt.
  const keyPath = join(tmpdir(), `cortex_tsauth_${process.pid}_${Date.now()}.key`)
  const scrub = () => { try { unlinkSync(keyPath) } catch { /* already gone */ } }
  let scrubbing = false
  const onSignal = (sig: string) => {
    if (scrubbing) return
    scrubbing = true
    scrub()
    process.exit(sig === 'SIGTERM' ? 143 : 130)
  }
  process.on('SIGINT', () => onSignal('SIGINT'))
  process.on('SIGTERM', () => onSignal('SIGTERM'))

  writeFileSync(keyPath, authKey, { mode: 0o600 })
  console.log('Bringing Tailscale up...')
  try {
    const proc = Bun.spawnSync(['tailscale', 'up', '--authkey', `file:${keyPath}`], {
      stdout: 'pipe',
      stderr: 'pipe',
    })

    if (proc.exitCode === 0) {
      console.log('Tailscale is up.')
      cmdStatus()
    } else {
      console.error(`ERROR: ${proc.stderr.toString().trim()}`)
      process.exit(1)
    }
  } finally {
    scrub()
  }
}

// --- Down ---

function cmdDown(): void {
  requireTailscale()
  const proc = Bun.spawnSync(['tailscale', 'down'], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode === 0) {
    console.log('Tailscale is down.')
  } else {
    console.error(`ERROR: ${proc.stderr.toString().trim()}`)
    process.exit(1)
  }
}

// --- IP ---

function cmdIp(hostname: string): void {
  requireTailscale()
  const proc = Bun.spawnSync(['tailscale', 'ip', hostname], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode === 0) {
    const ip = proc.stdout.toString().trim()
    console.log(`${hostname}: ${ip}`)
    console.log('\nUse this IP in your rclone SFTP remote config:')
    console.log(`  host = ${ip}`)
  } else {
    console.error(`ERROR: ${proc.stderr.toString().trim()}`)
    console.error('Is Tailscale up? Run: bun manifest/framework/scripts/integrations/tailscale.ts up')
    process.exit(1)
  }
}

// --- Peers ---

function cmdPeers(): void {
  requireTailscale()
  const proc = Bun.spawnSync(['tailscale', 'status', '--json'], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode !== 0) {
    console.error('Tailscale is not running. Run: bun manifest/framework/scripts/integrations/tailscale.ts up')
    process.exit(1)
  }

  let data: {
    Self?: { TailscaleIPs?: string[]; HostName?: string }
    Peer?: Record<string, { TailscaleIPs?: string[]; HostName?: string; Online?: boolean; OS?: string }>
  }
  try {
    data = JSON.parse(proc.stdout.toString())
  } catch {
    console.error('ERROR: Could not parse tailscale status output.')
    process.exit(1)
  }

  const peers = data.Peer ?? {}
  const selfNode = data.Self ?? {}

  console.log('# Tailnet peers\n')

  const selfIps = selfNode.TailscaleIPs ?? []
  const selfHost = selfNode.HostName ?? 'this device'
  const selfIp = selfIps[0] ?? 'unknown'
  console.log(`  ${selfIp}  ${selfHost}  (this device)`)

  for (const peer of Object.values(peers)) {
    const ips = peer.TailscaleIPs ?? []
    const ip = ips[0] ?? 'unknown'
    const hostname = peer.HostName ?? 'unknown'
    const online = peer.Online ? 'online' : 'offline'
    const osName = peer.OS ?? ''
    const osStr = osName ? ` [${osName}]` : ''
    console.log(`  ${ip}  ${hostname}${osStr}  (${online})`)
  }

  console.log('\nTo get IP for rclone: bun manifest/framework/scripts/integrations/tailscale.ts ip <hostname>')
}

// --- Main ---

async function main(): Promise<void> {
  const argv = process.argv.slice(2)
  if (argv.length === 0) {
    console.log('Usage: bun manifest/framework/scripts/integrations/tailscale.ts [--passphrase <p>] <subcommand>')
    console.log('Subcommands: auth, status, up, down, ip <hostname>, peers')
    process.exit(1)
  }

  // Pull out top-level --passphrase before subcommand
  let passphrase: string | undefined
  const rest: string[] = []
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--passphrase' && argv[i + 1]) {
      passphrase = argv[i + 1]
      i++
    } else {
      rest.push(argv[i])
    }
  }

  const cmd = rest[0]
  const cmdArgs = parseArgs(rest.slice(1))

  if (!cmd) {
    console.error('ERROR: No subcommand provided.')
    process.exit(1)
  }

  // Commands that don't need the vault
  if (cmd === 'status') { cmdStatus(); return }
  if (cmd === 'down') { cmdDown(); return }
  if (cmd === 'peers') { cmdPeers(); return }
  if (cmd === 'ip') {
    const hostname = cmdArgs['_cmd'] as string ?? rest[1]
    if (!hostname) { console.error('ERROR: ip requires <hostname>'); process.exit(1) }
    cmdIp(hostname)
    return
  }

  // Commands that need the vault
  if (!passphrase) passphrase = await promptPassphrase()

  if (cmd === 'auth') {
    await cmdAuth(passphrase)
  } else if (cmd === 'up') {
    await cmdUp(passphrase)
  } else {
    console.error(`ERROR: Unknown subcommand '${cmd}'.`)
    process.exit(1)
  }
}

main().catch(err => {
  console.error(`ERROR: ${(err as Error).message}`)
  process.exit(1)
})
